src/transformers/pipelines/text_to_speech.py

```diff
@@ -0,0 +1,173 @@
+        # for which there are no SpeechT5ForTextToSpeechWithHifiGanHead
+        self.is_speecht5 = self.model_type == "speecht5"
+
+        # check if the model only takes one speaker embeddings per generation
+        if self.model_type in ONLY_ONE_SPEAKER_EMBEDDINGS_LIST:
+            self.only_one_speaker_embeddings = True
+        else:
+            self.only_one_speaker_embeddings = False
+
+        if self.is_speecht5 and vocoder is None:
+            raise ValueError(
+                """vocoder is None but speecht5 is used.
+                              Try passing a repo_id or an instance of SpeechT5HifiGan."""
+            )
+        elif self.is_speecht5 and isinstance(vocoder, str):
+            vocoder = SpeechT5HifiGan.from_pretrained(vocoder).to(self.model.device)
+        elif self.is_speecht5 and not isinstance(vocoder, SpeechT5HifiGan):
+            raise ValueError(
+                """Must pass a valid vocoder to the TTSPipeline if speecht5 is used.
+                              Try passing a repo_id or an instance of SpeechT5HifiGan."""
+            )
+
+        self.processor = processor
+        self.vocoder = vocoder
+
+        if self.is_speecht5:
+            self.sampling_rate = vocoder.config.sampling_rate
+        elif sampling_rate is None:
+            # get sampling_rate from config and generation config
+            self.sampling_rate = None
+
+            config = self.model.config.to_dict()
+            gen_config = self.model.__dict__.get("generation_config", None)
+            if gen_config is not None:
+                config.update(gen_config.to_dict())
+
+            for sampling_rate_name in ["sample_rate", "sampling_rate"]:
+                sampling_rate = config.get(sampling_rate_name, None)
+                if sampling_rate is not None:
+                    self.sampling_rate = sampling_rate
+        else:
+            self.sampling_rate = sampling_rate
+
+    def preprocess(self, text, speaker_embeddings=None, **kwargs):
+        if self.is_speecht5:
+            inputs = self.processor(text=text, return_tensors="pt")
+            if speaker_embeddings is None:
+                embeddings_dataset = load_dataset("Matthijs/cmu-arctic-xvectors", split="validation")
+                speaker_embeddings = torch.tensor(embeddings_dataset[7305]["xvector"]).unsqueeze(0)
+
+            return {"input_ids": inputs["input_ids"], "speaker_embeddings": speaker_embeddings}
+
+        processor_args_dict = {}
+
+        SPEAKER_EMBEDDINGS_KEY_MAPPING.get(self.model_type, "speaker_embeddings")
+        if self.model.config.model_type == "bark":
+            # bark speaker embeddings is passed as voice_preset in its processor
+            processor_args_dict["voice_preset"] = speaker_embeddings
+        else:
+            processor_args_dict["speaker_embeddings"] = speaker_embeddings
+
+        output = self.processor(text, **processor_args_dict, **kwargs)
```

Comment:
Basically no to everything here.

Using `processor` is showing to be pretty bad here as you can see all the weird ifs everywhere.

```
return self.tokenizer(text, return_tensors="pt")
```

Should be pretty much it.

You cannot load arbitrary datasets (especially within a running pipeline).

At most `user_embeddings` can be sent by the user because iiuc there's no correct way to get it automatically and consistently from anywhere. (meaning it's now in sanitize_parameters, and you just receive it, and cannot create it here).

Reply:
Hi @Narsil , thanks for the exhaustive review!

In terms of TTS models, one of the clear advantages and one of the features most eagerly awaited by users could be the easy use of `speaker_embeddings` when the model in use allows it.

I totally understand now the need to stay consistent with the stateless pipeline logic, and thus not to introduce any `self.` objects and not to use `Processor`. But do you think there might be a way to stay consistent with the library and still uses some of the advantages of the `Processor` (i.e using  conditional objects) ?